crates/amaru/src/bin/amaru/cmd/convert_ledger_state.rs

```diff
@@ -0,0 +1,372 @@
+use clap::Parser;
+use std::path::{Path, PathBuf};
+use tokio::fs::{self};
+
+#[derive(Debug, Parser)]
+pub struct Args {
+    /// Path to the CBOR encoded ledger state snapshot as serialised by Haskell
+    /// node.
+    #[arg(long, value_name = "SNAPSHOT", verbatim_doc_comment, num_args(0..))]
+    snapshot: Vec<PathBuf>,
+
+    /// Directory to store converted snapshots into.
+    ///
+    /// Directory will be created if it does not exist, defaults to '.'.
+    #[arg(long, value_name = "DIR", verbatim_doc_comment)]
+    target_dir: Option<PathBuf>,
+
+    /// Network to convert snapshots for.
+    #[arg(long, value_name = "NETWORK_NAME", verbatim_doc_comment)]
+    network: NetworkName,
+}
+
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum Error {
+    #[error("Snapshot {0} does not exist")]
+    SnapshotDoesNotExist(PathBuf),
+    #[error("Snapshot {0} is not a file")]
+    SnapshotIsNotFile(PathBuf),
+}
+
+pub(crate) async fn run(args: Args) -> Result<(), Box<dyn std::error::Error>> {
+    let target_dir = args.target_dir.unwrap_or(PathBuf::from("."));
+    for snapshot in args.snapshot {
+        convert_one_snapshot_file(&target_dir, &snapshot, &args.network).await?;
+    }
+
+    Ok(())
+}
+
+async fn convert_one_snapshot_file(
+    target_dir: &Path,
+    snapshot: &PathBuf,
+    network: &NetworkName,
+) -> Result<PathBuf, Box<dyn std::error::Error>> {
+    if !snapshot.exists() {
+        return Err(Box::new(Error::SnapshotDoesNotExist(snapshot.clone())));
+    }
+    if !snapshot.is_file() {
+        return Err(Box::new(Error::SnapshotIsNotFile(snapshot.clone())));
+    }
+
+    fs::create_dir_all(target_dir).await?;
+
+    convert_snapshot_to(snapshot, target_dir, network).await
+}
+
+async fn convert_snapshot_to(
+    snapshot: &PathBuf,
+    target_dir: &Path,
+    network: &NetworkName,
+) -> Result<PathBuf, Box<dyn std::error::Error>> {
+    let bytes = fs::read(snapshot).await?;
+    let mut d = cbor::Decoder::new(bytes.as_slice());
+
+    d.array()?;
+
+    // version
+    // https://github.com/abailly/ouroboros-consensus/blob/1508638f832772d21874e18e48b908fcb791cd49/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/Versioned.hs#L95
+    d.skip()?;
+
+    // ext ledger state
+    // https://github.com/abailly/ouroboros-consensus/blob/1508638f832772d21874e18e48b908fcb791cd49/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Extended.hs#L232
+    d.array()?;
+
+    // ledger state
+    d.array()?;
+
+    // HardForkCombinator telescope encoding.
+    // encodes the era history. There are 6 eras before conway.
+    // FIXME: pass the current Era to know how many skips to do
+    let mut eras: Vec<Summary> = decode_eras(&mut d, network)?;
+
+    // current era lower bound
+    d.array()?;
+    let start: Bound = d.decode()?;
+    eras.push(Summary {
+        start,
+        end: None,
+        params: EraParams {
+            epoch_size_slots: network.default_epoch_size_in_slots(),
```

Comment:
Shouldn't this be inferred from the `eras` instead of being hardwired to the network? It probably doesn't matter for well-known networks, but for devnets which can have arbitrary globals, that might not end well?